File: http_server.go
package main

import (
	"fmt"
	"log"
	"net"
	"net/http"
)

func init() {
	log.Println("Initializing HTTP server...")
}

func main() {
	if err := startServer(); err != nil {
		log.Fatal("Server error:", err)
	}
}

func startServer() error {
	log.Println("Starting HTTP server on port 8080...")

	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Welcome to the Go HTTP Server!")
	})
	// Define additional routes
	// You can add more routes here as needed
	// Example routes:
	// http.HandleFunc("/hello", func(w http.ResponseWriter, r *http.Request) {
	// 	fmt.Fprintf(w, "Hello, Go!")
	// })

	http.HandleFunc("/hello", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Hello, Go!")
	})
	// Another example route
	// You can add more routes here as needed
	// Example route:
	// http.HandleFunc("/goodbye", func(w http.ResponseWriter, r *http.Request) {
	// 	fmt.Fprintf(w, "Goodbye, World!")
	// })

	http.HandleFunc("/goodbye", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Goodbye, World!")
	})

	http.HandleFunc("/greet", func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintf(w, "Greetings, World!")
	})

	err := http.ListenAndServe(":8080", nil)
	// Log the error if the server fails to start
	// This will help in debugging if there are issues with the server setup
	// The ListenAndServe function will block and run the server until it encounters an error

	if err != nil {
		log.Fatal("Error starting server:", err)
	}
	// If the server starts successfully, it will run indefinitely until an error occurs
	// The server will handle incoming requests and respond according to the defined routes
	// The main function will not exit until the server encounters an error, at which point it will log the error and terminate the program
	// The server will continue to run and handle requests until it is stopped or encounters an error
	// You can stop the server by sending an interrupt signal (e.g., Ctrl+C) in the terminal where it is running
	// The server will then log the shutdown message and exit gracefully
	log.Println("HTTP server is running on port 8080...")


	return nil
}


// how to write html in go

// To write HTML in Go, you can use the `html/template` package, which provides a way to generate HTML output safely. Here's an example of how to use it:


func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		tmpl := template.Must(template.New("index").Parse(`
			<!DOCTYPE html>
			<html>
			<head>
				<title>Go HTML Template</title>
			</head>
			<body>
				<h1>Welcome to Go HTML Template!</h1>
				<p>This is an example of how to write HTML in Go.</p>
			</body>
			</html>
		`))

		if err := tmpl.Execute(w, nil); err != nil {
			log.Println("Error executing template:", err)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
		}
	})

	log.Println("Starting server on :8080")
	if err := http.ListenAndServe(":8080", nil); err != nil {
		log.Fatal("ListenAndServe:", err)
	}
}

// In this example, we define an HTML template as a string and use the `template.Must` function to parse it. 
// The `Execute` method is then called to write the generated HTML to the response writer. 
// You can customize the HTML template as needed, and you can also pass data to the template for dynamic content generation.
// When you run this code, it will start an HTTP server on port 8080.
// When you access `http://localhost:8080` in your web browser, you will see the rendered HTML page with the message "Welcome to Go HTML Template!" and a paragraph explaining that this is an example of how to write HTML in Go.

// how to write javascript in go	
// To write JavaScript in Go, you can include it within your HTML templates. Here's an example of how to do this using the `html/template` package:

package main

import (
	"html/template"
	"log"
	"net/http"
)

func main() {
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		tmpl := template.Must(template.New("index").Parse(`
			<!DOCTYPE html>
			<html>
			<head>
				<title>Go HTML Template with JavaScript</title>
			</head>
			<body>
				<h1>Welcome to Go HTML Template with JavaScript!</h1>
				<p>This is an example of how to write JavaScript in Go.</p>
				<button onclick="showMessage()">Click Me</button>

				<script>
					function showMessage() {
						alert("Hello from JavaScript!");
					}
				</script>
			</body>
			</html>
		`))

		if err := tmpl.Execute(w, nil); err != nil {
			log.Println("Error executing template:", err)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
		}
	})

	log.Println("Starting server on :8080")
	if err := http.ListenAndServe(":8080", nil); err != nil {
		log.Fatal("ListenAndServe:", err)
	}
}

// In this example, we include a simple JavaScript function `showMessage()` that displays an alert when the button is clicked. 
// The JavaScript code is embedded directly within the HTML template. 
// When you run this code and access `http://localhost:8080`, you will see a button on the page. 
// When you click the button, it will trigger the JavaScript function and display an alert with the message "Hello from JavaScript!"
// You can customize the JavaScript code as needed, and you can also pass data from Go to the template to create dynamic JavaScript content.

// lets write a simple http server

package main

import (
	"fmt"
	"net"
)

func handleConnection_advanced(conn net.Conn) {
	defer conn.Close()
	fmt.Println("New connection accepted...")

	// Read the request (not implemented in this example)
	// You can read the request data from the connection and process it as needed

	// Write a simple HTTP response
	response := "HTTP/1.1 200 OK\r\n" +
		"Content-Type: text/plain\r\n" +
		"\r\n" +
		"Hello, World!"
	conn.Write([]byte(response))
	// The connection will be closed after the response is sent
	// You can also implement more complex request handling and response generation as needed
	// This is a very basic example of an HTTP server. 
	// In a real application, you would need to parse the incoming HTTP request, handle different routes, and generate appropriate responses based on the request data.
}

func handleConnection_classic(conn net.Conn) {
	defer conn.Close()
	message, _ := bufio.NewReader(conn).ReadString('\n')
	fmt.Printf("Received message: %s", message)
	conn.Write([]byte("Message received: " + message))
}

// In this example, we create a simple HTTP server that listens on port 8080. 
// When a new connection is accepted, it spawns a new goroutine to handle the connection. 
// The `handleConnection` function reads the request (not implemented in this example) and writes a simple HTTP response back to the client. 
// When you run this code and access `http://localhost:8080`, you will receive a plain text response with the message "Hello, World!"

func main() {
	// listener, _ := net.Listen("tcp", ":8080")
	// defer listener.Close()
	// fmt.Println("Server is listening on port 8080...")

	// for {
	// 	conn, _ := listener.Accept()
	// 	go handleConnection(conn)
	// }
	// The above code is a simple TCP server that listens for incoming connections on port 8080. 
	// When a new connection is accepted, it spawns a new goroutine to handle the connection using the `handleConnection` function. 
	// The server will continue to run indefinitely, accepting new connections and handling them concurrently.
	// You can test the server by connecting to it using a web browser or a tool like `curl` and sending an HTTP request. 
	// The server will respond with a simple HTTP response containing the message "Hello, World!"

	conn, _ := net.Dial("tcp", "127.0.0.1:8080")
	defer conn.Close()
	fmt.Fprintf(conn, "Hello, Server!\n")
	response, _ := bufio.NewReader(conn).ReadString('\n')
	fmt.Printf("Response from server: %s", response)
	// The above code is a simple TCP client that connects to the server at `8080` and sends a message "Hello, Server!" to the server. 
	// It then reads the response from the server and prints it to the console. 
	// You can run this client code in a separate program or goroutine while the server is running to test the communication between the client and server.
}

File: http_server.cpp
// how to write http server in c++14To write a simple HTTP server in C++14, you can use the Boost.
// Asio library, which provides a cross-platform way to handle network programming. 
// Below is an example of a basic HTTP server that listens on port 8080 and responds with a simple "Hello, World!" message to any incoming HTTP request.

#include <boost/asio.hpp>

// working compiltion:

// g++ -std=c++14 http_server.cpp -o http_server -I/opt/homebrew/include -L/opt/homebrew/lib -lpthread && ./http_server

#include <iostream>
#include <string>

// you change io_service to io_context in boost 1.66 and later versions, so make sure to use the correct version of boost for your project.
// cool 
// how to run this code?
// To run this code, follow these steps:
// 1. Make sure you have Boost installed on your system. You can install it using a package manager like Homebrew on macOS (brew install boost) or using your system's package manager on Linux.
// 2. Save the code to a file named hello.cpp.
// 3. Open a terminal and navigate to the directory where you saved hello.cpp.
// 4. Compile the code using the following command (adjusting the include path if necessary):
//    g++ -std=c++14 http_server.cpp -o http_server -I/opt/homebrew/include -L/opt/homebrew/lib -lpthread
// 5. Run the compiled program:
//    ./hello

// or just g++ -std=c++14 http_server.cpp -o http_server -I/opt/homebrew/include -L/opt/homebrew/lib -lpthread && ./http_server

// 6. The server will start and listen on port 8080. You can test it by opening a web browser and navigating to http://localhost:8080, or by using a tool like curl:
//    curl http://localhost:8080


int main() {
	try {
		boost::asio::io_context io_context;

		// Create an acceptor to listen for incoming connections on port 8080
		
		boost::asio::ip::tcp::acceptor acceptor(io_context, boost::asio::ip::tcp::endpoint(boost::asio::ip::tcp::v4(), 8080));
		
		// acceptor - This is an object that listens for incoming TCP connections on a specified endpoint (in this case, port 8080).
		// boost::asio::ip::tcp::endpoint - This represents an endpoint for TCP communication, which consists of an IP address and a port number. 
		// Here, we are using boost::asio::ip::tcp::v4() to specify that we want to listen on all available IPv4 addresses.

		// io_context - This is the main I/O context object that manages asynchronous operations. It is used to run the event loop that handles incoming connections and other asynchronous tasks.

		// endpoint - This is a specific address and port combination that the server will listen on. In this case, we are using boost::asio::ip::tcp::v4() to specify that we want to listen on all available IPv4 addresses, and port 8080.

		// v4() - This is a function that returns an object representing the IPv4 protocol. It is used to specify that we want to listen for TCP connections over IPv4.

		std::cout << "HTTP server is running on port 8080..." << std::endl;

		while (true) {
			// Create a socket to hold the incoming connection
			boost::asio::ip::tcp::socket socket(io_context);
			acceptor.accept(socket);

			// Read the HTTP request (not fully implemented for simplicity)
			boost::asio::streambuf request;
			boost::asio::read_until(socket, request, "\r\n\r\n");

			// Prepare the HTTP response
			std::string response = 
				"HTTP/1.1 200 OK\r\n"
				"Content-Type: text/plain\r\n"
				"Content-Length: 13\r\n"
				"\r\n"
				"Hello, World!";

			// Write the response back to the client
			boost::asio::write(socket, boost::asio::buffer(response));
		}
	} catch (std::exception& e) {
		std::cerr << "Exception: " << e.what() << std::endl;
	}

	return 0;
}